src/safe-event-emitter-provider.ts

```diff
@@ -0,0 +1,62 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import type { JsonRpcEngine, JsonRpcRequest } from 'json-rpc-engine';
+
+/**
+ * An Ethereum provider.
+ *
+ * This provider loosly follows conventions that pre-date EIP-1193. It
+ * is not compliant with any Ethereum provider standard.
+ */
+export class SafeEventEmitterProvider extends SafeEventEmitter {
+  #engine: JsonRpcEngine;
+
+  /**
+   * Construct a SafeEventEmitterProvider from a JSON-RPC engine.
+   *
+   * @param options - Options.
+   * @param options.engine - The JSON-RPC engine used to process requests.
+   */
+  constructor({ engine }: { engine: JsonRpcEngine }) {
+    super();
+    this.#engine = engine;
+
+    if (engine.on) {
+      engine.on('notification', (message: string) => {
+        this.emit('data', null, message);
+      });
+    }
+  }
+
+  /**
+   * Send a provider request asynchronously.
+   *
+   * @param req - The request to send.
+   * @param callback - A function that is called upon the success or failure of the request.
+   */
+  sendAsync = (
+    req: JsonRpcRequest<unknown>,
+    callback: (error: unknown, providerRes?: any) => void,
+  ) => {
+    this.#engine.handle(req, callback);
+  };
+
+  /**
+   * Send a provider request asynchronously.
```

Comment:
```suggestion
   * Send a provider request synchronously.
```

Reply:
This version is still asynchronous. The name comes from a time when `send` was partially synchronous and partially async. `sendAsync` was different in that it was always async. But we have not had any sync support in this method for some time now.